### DISCUSSION TITLE
Collision Resolution for Kinematic bodies

### DISCUSSION THREAD

**Discussion by spellberry** [jrouwe/JoltPhysics#860] [General]
I am working on a project where you have a controllable character (currently using the "CharacterVirtual" class from Jolt) which can shoot projectiles and cast AOE abilities. I managed to get collision detection (using the "ContactListener") and to differentiate if a projectile hits terrain or another character. My questions are:

- How would I resolve the collision between the abilities and the enemy character? I can't do it in the "OnContactAdded" function because it freezes the program so I feel like it probably shouldn't happen there.
- How would I go about making an AOE type of ability? At the moment I have a scaled shape, which is a cylinder, and it looks a bit like a hockey puck, but I'd want it to be just a flat circle with which other objects that are "CharacterVirtual" can't collide with it. Should I just make it a sensor?

**Comment by jrouwe** [comment]
> How would I resolve the collision between the abilities and the enemy character? I can't do it in the "OnContactAdded" function because it freezes the program so I feel like it probably shouldn't happen there.

In the OnContactAdded callback all bodies are locked, it sounds like the the program is freezing because of locking a mutex that is already locked on the same thread, but without more details it is hard to be sure. Have you tried using the non-locking BodyInterface inside the OnContactAdded callback? (note that inside that callback you should not modify physics state as this part of the physics update is multi threaded)

> How would I go about making an AOE type of ability? At the moment I have a scaled shape, which is a cylinder, and it looks a bit like a hockey puck, but I'd want it to be just a flat circle with which other objects that are "CharacterVirtual" can't collide with it. Should I just make it a sensor?

Circles are not supported as collision primitive, but you can make a short cylinder. I'm not sure if an AOE needs a body/sensor at all, you can just do a collision check through NarrowPhaseQuery when you want to deal damage / do whatever the AOE does. When you make it a sensor then that is pretty much what happens internally anyway.

**Reply by spellberry** [reply]
> Circles are not supported as collision primitive, but you can make a short cylinder. I'm not sure if an AOE needs a body/sensor at all, you can just do a collision check through NarrowPhaseQuery when you want to deal damage / do whatever the AOE does. When you make it a sensor then that is pretty much what happens internally anyway.

How exactly would you do this? Right now, I have a Jolt Body for the AOE (the short cylinder that you said), and another attached to it as a sensor. Since it will go through the NarrowPhaseQuery, should I just have one body as a sensor? The main issue is that I don't want characters to collide it as if it was a wall (they are doing it as of now).

**Reply by jrouwe** [reply]
> How exactly would you do this? Right now, I have a Jolt Body for the AOE (the short cylinder that you said), and another attached to it as a sensor. Since it will go through the NarrowPhaseQuery, should I just have one body as a sensor? The main issue is that I don't want characters to collide it as if it was a wall (they are doing it as of now).

Why not only create a Shape (so no body and no sensor) and use NarrowPhaseQuery::CollideShape to test collision with the world. That way it won't be in the world at all so nothing else can see/collide with it.
